### ISSUE TITLE
User-defined heatmap value range

### ISSUE THREAD

**Issue by safay** [ewels/MultiQC#1427]
**Is your feature request related to a problem? Please describe.**

Lab users would like to be able to view differences within a variable range of values within a heatmap. For example, when there are some samples with very high values (e.g., 30,000, see attached plot) but for the samples we are particularly interested in we need to distinguish between values within a range of, say, 1-500, the higher value samples burn out the signal. Even if the user hides the samples with high values, the colors of the heatmap do not change.

**Describe the solution you'd like**

A brute force solution would be to implement a hard cutoff/limit for the heatmap values, something like the "xmax" config for x/y plots. If this exists today I could not find it.
A heavier lift, more UI-based, solution would be to have something like a slider on the sidebar that could dynamically re-render the heatmap for a given range.

**Describe alternatives you've considered**

See above about using a config setting for heatmaps analogous to xmax for x/y plots.
I've considered rolling my own solution in matplotlib outside MultiQC, but I'm adding this feature request here because it seems like it might be something others might want.
I could add a new part to the analysis pipeline to separate out low-signal samples and create a separate heatmap for those samples.

**Additional context**

I'm willing to have a stab at making a PR, especially if I can get some guidance.

![image](https://user-images.githubusercontent.com/2934664/118065948-af059180-b352-11eb-9806-bc9197fd2297.png)

**Comment by ewels** [comment]
Hi @safay,

I think that the plot does have the option to set a hard cutoff, it's just called `max` instead of `xmax` though: https://multiqc.info/docs/#heatmaps

```python
pconfig = {
    'min': None,                   # Minimum value (default: auto)
    'max': None,                   # Maximum value (default: auto)
}
```

You can overwrite any plot config at run time as follows: https://multiqc.info/docs/#customising-plots

> Even if the user hides the samples with high values, the colors of the heatmap do not change.

Looking at the code, this is deliberate behaviour. I seem to remember doing that in the early days as it was quite confusing to have the scale / meaning of the colours change each time you filtered.

> A heavier lift, more UI-based, solution would be to have something like a slider on the sidebar that could dynamically re-render the heatmap for a given range.

This sounds great, but would not be trivial to implement.

Phil

**Comment by ewels** [comment]
ok you made me curious as I agree that the sliders would be a really nice feature. So I started playing and came up with this:

![heatmap-sliders](https://user-images.githubusercontent.com/465550/118244927-6a880c00-b4a0-11eb-9ad9-40f1379f924c.gif)

Added in 4abbfe5c45f91b4abfbfe3080b505207acbb5fdf - if you could please install the `dev` version of MultiQC and give it a try that would be fab 👍🏻 

Phil

**Comment by safay** [comment]
@ewels Thank you! This is great. Also, thanks for pointing me to the right place in the docs to find parameters for plot configs. 
Trying it now - will report back!

**Comment by safay** [comment]
@ewels This is brilliant. I know beggars can't be choosers, but _if_ it would be trivial to add, a field to enter min/max would be nice.
![image](https://user-images.githubusercontent.com/2934664/118339962-40017780-b4cf-11eb-9a61-f082dff549d1.png)

**Comment by ewels** [comment]
Yes I was looking at this again last night before I went to bed and was thinking exactly that - agree that it would be useful (as shown by your screenshot with small useful ranges) and should be easy enough to add. Will have a look 👍🏻

**Comment by ewels** [comment]
ok, added in 26177e9 - please give that a whirl and see how it looks 👍🏻 

![May-15-2021 14-01-23](https://user-images.githubusercontent.com/465550/118360101-67ffe200-b586-11eb-94c8-96018a606594.gif)

**Comment by safay** [comment]
@ewels I tested it today and it's lovely, that works really really well. I am very grateful!

![image](https://user-images.githubusercontent.com/2934664/118568653-278e9880-b72d-11eb-82c0-b2d49853e871.png)

![image](https://user-images.githubusercontent.com/2934664/118568693-3a08d200-b72d-11eb-83ba-00d2a00cf127.png)

**Comment by safay** [comment]
New feature implemented in `26177e9`

**Comment by ewels** [comment]
Brilliant, thanks for the suggestion and prompt!
